### PR TITLE
[BEAM-14038] Auto-startup for Python expansion service.

### PR DESCRIPTION
**Pull request by robertwb**
This adds a script and some java code for automatically starting a Beam Python service.

It may be useful for Go, Typescript, etc. as well. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @ihji

**Comment by robertwb** [comment]
Failure in org.apache.beam.sdk.transforms.ParDoLifecycleTest.testTeardownCalledAfterExceptionInFinishBundleStateful

**Comment by robertwb** [comment]
Run Java PreCommit

**Comment by robertwb** [comment]
Flake due to BEAM-8101

**Comment by robertwb** [comment]
Run Java PreCommit

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        PythonService.waitForPort("localhost", port, 5000);
+        PTransform<PInput, PCollectionTuple> transform =
+            External.<PInput, Object>of(
+                    "beam:transforms:python:fully_qualified_named",
+                    payload.toByteArray(),
+                    "localhost:" + port)
+                .withMultiOutputs();
+        PCollectionTuple outputs;
+        if (input instanceof PCollection) {
+          outputs = (PCollectionTuple) ((PCollection) input).apply(transform);
```
Comment (ihji) [inline sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:89]: Do we need this explicit casting to `(PCollectionTuple)` ?

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonService.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+  public AutoCloseable start() throws IOException, InterruptedException {
+    File bootstrapScript = File.createTempFile("bootstrap_beam_venv", ".py");
+    bootstrapScript.deleteOnExit();
+    try (FileOutputStream fout = new FileOutputStream(bootstrapScript.getAbsolutePath())) {
+      ByteStreams.copy(getClass().getResourceAsStream("bootstrap_beam_venv.py"), fout);
+    }
+    List<String> bootstrapCommand = ImmutableList.of("python", bootstrapScript.getAbsolutePath());
+    LOG.info("Running bootstrap command " + bootstrapCommand);
+    Process bootstrap =
+        new ProcessBuilder("python", bootstrapScript.getAbsolutePath())
```
Comment (ihji) [inline sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonService.java:62]: Isn't it simpler to run `bootstrap_beam_venv.py` not by creating a temporary copy but just passing the script text resource to python executable via stdin pipe?

File: sdks/java/extensions/python/src/main/resources/org/apache/beam/sdk/extensions/python/bootstrap_beam_venv.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    if sys.version_info < (3, ):
+        # Run this script with Python 3.
+        os.execlp('python3', 'python3', *sys.argv)
+        return  # In case windows returns...
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument('--python_version', help="Python major version.")
+    parser.add_argument('--beam_version',
+                        help="Beam version.",
+                        default="2.36.0")
```
Comment (ihji) [inline sdks/java/extensions/python/src/main/resources/org/apache/beam/sdk/extensions/python/bootstrap_beam_venv.py:45]: Is there any way to avoid hard-coded Beam version? Do we need to update this version string every time we release a new version of Beam?

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+              "--port",
+              "" + port,
+              "--fully_qualified_name_glob",
+              "*");
+      Schema payloadSchema =
+          Schema.of(
+              Schema.Field.of("constructor", Schema.FieldType.STRING),
+              Schema.Field.of("args", Schema.FieldType.row(args.getSchema())),
+              Schema.Field.of("kwargs", Schema.FieldType.row(kwargs.getSchema())));
+      payloadSchema.setUUID(UUID.randomUUID());
```
Comment (ihji) [inline sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:69]: Just out of curiosity, why do we need to set UUID for the schema?

**Review by ihji — COMMENTED** [review]
Thanks, looks good.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 1 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:89]: Yes, unfortunately. It seems java struggles with "paritally-raw" types.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonService.java (hunk 2 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/PythonService.java:62]: Good question, I actually tried to do it this way originally. The problem is that the `os.execlp` bootstrap from Python 2 to Python 3 doesn't work if it isn't run as a file (as the stdin is already partially consumed). There are still too many systems out there that, despite having Python 3, map `python` to Python 2.x. (An alternative would be to make the full script Python 2 compatible, but that would cause ongoing pain, especially if one can't count on any dependencies like six...)

File: sdks/java/extensions/python/src/main/resources/org/apache/beam/sdk/extensions/python/bootstrap_beam_venv.py (hunk 3 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/resources/org/apache/beam/sdk/extensions/python/bootstrap_beam_venv.py:45]: We could consider making this a required argument. (Note that it's not an error to mix and match Beam versions, as long as what you're counting on is available, though it's generally best to have the latest.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 4 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:69]: Python has a schema cache (e.g. so it's not constantly re-creating types for a dynamic schema). Results in some very surprising behavior if this is not set.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 1 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:89]: OK, I was able to work around this by using `PCollection<?>`.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.PInput;
+import org.apache.beam.sdk.values.POutput;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.vendor.grpc.v1p43p2.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Iterables;
+
+/** Wrapper for invoking external Python transforms. */
+public class ExternalPythonTransform<InputT extends PInput, OutputT extends POutput>
```
Comment (chamikaramj) [inline sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:40]: We want this to be the main API for Java cross-language transforms, right ? So probably we should advertise in the PR description.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+    extends PTransform<InputT, OutputT> {
+  private final String fullyQualifiedName;
+  private final Row args;
+  private final Row kwargs;
+
+  public ExternalPythonTransform(String fullyQualifiedName, Row args, Row kwargs) {
```
Comment (chamikaramj) [inline sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:46]: How about adding following that wraps this transform:
External.of (String transformName, Map parameters, String endpoint);

And may be deprecate existing:
External.of(String urn, byte[] payload, String endpoint)

(this can be a separate PR)

**Review by chamikaramj — COMMENTED** [review]
Thanks.

File: sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransformTest.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+public class ExternalPythonTransformTest implements Serializable {
+  @Test
+  public void trivialPythonTransform() {
+    Pipeline p = Pipeline.create();
+    PCollection<String> output =
+        p.apply(Create.of(KV.of("A", "x"), KV.of("A", "y"), KV.of("B", "z")))
+            .apply(
+                new ExternalPythonTransform<
+                    PCollection<KV<String, String>>, PCollection<KV<String, Iterable<String>>>>(
+                    "apache_beam.GroupByKey", Row.nullRow(Schema.of()), Row.nullRow(Schema.of())))
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransformTest.java:44]: The Java DirectRunner can't run cross-language pipelines. I used GBK because I wanted something simple and obvious--eventually we'll hopefully have a richer set of transforms to invoke.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 6 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:46]: I agree that we need a nicer API, which should probably have a 1-pager and be a separate PR. (To seed this, we should have a RowBuilder that starts with the empty row and can be built with `withField(name, value)` with several overloads for different value types that builds up the schema dynamically. I don't know that a Map<String, Object> parameters is the most java-natural, but it could be allowed as an option as well.) I'm not convinced it makes sense to add a back-reference from External to ExternalPython.

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 5 above)
Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:40]: See below.

**Comment by robertwb** [comment]
Run Java PreCommit

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 6 above)
Comment (chamikaramj) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:46]: Adding a Row builder (on top of this) seems relatively straightforward. See https://github.com/apache/beam/pull/17101

I think it will be confusing to add another main External transform API (in addition to External.of and ExternalPythonTransform) so probably ExternalPythonTransform should be the main API to be used along with the Row builder.

Comment (robertwb) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:46]: ExternalPythonTransform only makes sense for Python transforms, as it starts up a Python expansion service (rather than require one be provided) and expects a fqn + positional arg + kwargs rather than a urn + row (though one could argue row is basically kwargs).

Comment (chamikaramj) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:46]: Yeah, sorry, I meant "main API for using Python transforms from Java".

**Comment by robertwb** [comment]
Run Java PreCommit

**Comment by robertwb** [comment]
Run Java PreCommit

File: sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java (hunk 6 above)
Comment (chamikaramj) [reply sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/ExternalPythonTransform.java:46]: Here's a short doc that gives an updated API: https://docs.google.com/document/d/164rn4otqtQA-8QoLlaoGunUDjLSBBvS2DuR0bn344MQ/edit?usp=sharing

**Comment by robertwb** [comment]
Failure in org.apache.beam.sdk.io.hadoop.format.HadoopFormatIOElasticTest.classMethod

**Comment by lukecwik** [comment]
ExternalPythonTransformTest is extremely flaky, for example: https://ci-beam.apache.org/job/beam_PreCommit_Java_Phrase/4806/testReport/junit/org.apache.beam.sdk.extensions.python/ExternalPythonTransformTest/trivialPythonTransform/

Filed https://issues.apache.org/jira/browse/BEAM-14148
